[SPARK-24495][SQL] EnsureRequirement returns wrong plan when reordering equal keys

mgaido91 opened the pull request:
## What changes were proposed in this pull request?

`EnsureRequirement` in its `reorder` method currently assumes that the same key appears only once in the join condition. This of course might not be the case, and when it is not satisfied, it returns a wrong plan which produces a wrong result of the query.

## How was this patch tested?

added UT

mgaido91 committed:
[SPARK-24495][SQL] EnsureRequirement returns worng plan when reordering equal keys

SparkQA commented:
**[Test build #91666 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91666/testReport)** for PR 21529 at commit [`06858cd`](https://github.com/apache/spark/commit/06858cd5a7e1b11cd5fd1edb206296d75550a8d4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3904/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/13/
Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/exchange/EnsureRequirements.scala:
```diff
@@ -227,9 +228,16 @@ case class EnsureRequirements(conf: SQLConf) extends Rule[SparkPlan] {
       currentOrderOfKeys: Seq[Expression]): (Seq[Expression], Seq[Expression]) = {
     val leftKeysBuffer = ArrayBuffer[Expression]()
     val rightKeysBuffer = ArrayBuffer[Expression]()
+    val alreadyUsedIndexes = mutable.Set[Int]()
```
nit: maybe `pickedIndexes`?

cloud-fan commented:
good catch! thanks!

SparkQA commented:
**[Test build #91666 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91666/testReport)** for PR 21529 at commit [`06858cd`](https://github.com/apache/spark/commit/06858cd5a7e1b11cd5fd1edb206296d75550a8d4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91666/
Test PASSed.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:
```diff
@@ -679,6 +679,17 @@ class PlannerSuite extends SharedSQLContext {
+    withSQLConf(("spark.sql.shuffle.partitions", "1"),
+      ("spark.sql.constraintPropagation.enabled", "false"),
+      ("spark.sql.autoBroadcastJoinThreshold", "-1")) {
+      val df1 = spark.range(100)
```
we should make sure range has more than one partitions, otherwise we can't reproduce the bug.

cloud-fan replied:
I think this bug also applies to broadcast join(if the non-broadcast table is bucketed) and shuffle hash join. Can we also add tests for them?

cloud-fan commented:
I was surprised that this bug is present even if the table is not bucketed, and then I found out another problem in the code: `reorderJoinPredicates` transform the plan again. `reorderJoinPredicates` is called during plan transformation and we should not double transform the plan.

This makes the bug has a much larger impact: when we do a shuffle join, we add shuffle exchange to both of the join sides, which then triggers `reorderJoinPredicates` because of the double transformation problem.

@mgaido91 let's also remove the `transformUp` in `reorderJoinPredicates`

cloud-fan commented:
one followup: we should improve our golden file sql test, to run same queries with different configs. This is pretty important for the join tests, otherwise we only test broadcast join.

mgaido91 committed:
address comments

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/42/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3932/
Test PASSed.

mgaido91 commented:
thanks for your review @cloud-fan. Nice catch on the `transformUp`! I addressed all your comments. 

As far as the followup is regarded, we should decide if we want to support the possibility that with different configs we get different results or not. In the second case we have more flexibility, but we have also to be much more careful in order not to introduce bugs when different configs should not produce different results. Moreover, we would also have many more golden files...
What do you think?

SparkQA commented:
**[Test build #91705 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91705/testReport)** for PR 21529 at commit [`341f1b2`](https://github.com/apache/spark/commit/341f1b28572df079d5ed6868c6983610d2bd951f).

SparkQA commented:
**[Test build #91705 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91705/testReport)** for PR 21529 at commit [`341f1b2`](https://github.com/apache/spark/commit/341f1b28572df079d5ed6868c6983610d2bd951f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91705/
Test FAILed.

mgaido91 commented:
retest this please

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3936/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/46/
Test PASSed.

SparkQA commented:
**[Test build #91710 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91710/testReport)** for PR 21529 at commit [`341f1b2`](https://github.com/apache/spark/commit/341f1b28572df079d5ed6868c6983610d2bd951f).

SparkQA commented:
**[Test build #91710 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91710/testReport)** for PR 21529 at commit [`341f1b2`](https://github.com/apache/spark/commit/341f1b28572df079d5ed6868c6983610d2bd951f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91710/
Test FAILed.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:
```diff
@@ -679,6 +679,17 @@ class PlannerSuite extends SharedSQLContext {
     assert(rangeExecInZeroPartition.head.outputPartitioning == UnknownPartitioning(0))
   }
+
+  test("SPARK-24495: EnsureRequirements can return wrong plan when reusing the same key in join") {
```
This is actually an end-to-end test(checking result not plan), I think it's better to put it in `SQLQuerySuite`

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:
```diff
@@ -679,6 +679,17 @@ class PlannerSuite extends SharedSQLContext {
   }
+
+  test("SPARK-24495: EnsureRequirements can return wrong plan when reusing the same key in join") {
+    withSQLConf(("spark.sql.shuffle.partitions", "1"),
```
let's not use hard coded config name, use `SQLConf.SHUFFLE_PARTITIONS` instead.

a nit: we usually use `a -> b, c -> d, ...` to specify config pairs.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:
```diff
@@ -679,6 +679,17 @@ class PlannerSuite extends SharedSQLContext {
+    withSQLConf(("spark.sql.shuffle.partitions", "1"),
+      ("spark.sql.constraintPropagation.enabled", "false"),
+      ("spark.sql.autoBroadcastJoinThreshold", "-1")) {
+      val df1 = spark.range(100).repartition(2, $"id", $"id")
```
a simpler way `spark.range(0, 100, 1, numPartitions = 2)`

mgaido91 replied:
this way would not be ok, as we would have a `RangePartitioning` while the issue appears only with `HashPartitioning`

cloud-fan replied:
no, the issue can happen with range partition(because of the double transformation issue), the code in the ticket can reproduce the bug and it has no hash partitioning.

mgaido91 replied:
yes, but if we remove the transformUp as you correctly suggested, if we do not use HashPartitioning we do not test the proper behavior of the `reorder` method. So I think that using the HashPartitioning is a better test for this change. What do you think?

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:
```diff
@@ -679,6 +679,17 @@ class PlannerSuite extends SharedSQLContext {
+      ("spark.sql.constraintPropagation.enabled", "false"),
+      ("spark.sql.autoBroadcastJoinThreshold", "-1")) {
+      val df1 = spark.range(100).repartition(2, $"id", $"id")
+      val df2 = spark.range(100).select(($"id" * 2).as("b1"), (- $"id").as("b2"))
```
`($"id" * 2).as("b1")` -> `$"id".as("b1")`, to minimize the test.

cloud-fan commented:
I think we need to support both. testing different physical operators needs same result, testing something like type coercion mode needs different result. Anyway let's discuss it in the followup.

gatorsmile commented:
@mgaido91 Could you help improve the test coverage of joins `org.apache.spark.sql.JoinSuite`?  Due to the incomplete test case coverage, we did not discover this at the very beginning. We need to add more test cases to cover different join algorithms.

mgaido91 commented:
> I think we need to support both. testing different physical operators needs same result, testing something like type coercion mode needs different result. Anyway let's discuss it in the followup.

ok @cloud-fan, I'll try and send a proposal in the next days.

> @mgaido91 Could you help improve the test coverage of joins org.apache.spark.sql.JoinSuite? Due to the incomplete test case coverage, we did not discover this at the very beginning. We need to add more test cases to cover different join algorithms.

Sure, @gatorsmile, I am happy to. Do you mean running the existing tests for every type of join or do you have something different in mind? Thanks.

cloud-fan commented:
I think for this PR, apart from the end-to-end test for checking result, we should also have a unit test in `PlannerSuite` to cover this case. In the followup, we can add the config support for the golden file tests and add more test cases.

mgaido91 committed:
address comments

SparkQA commented:
**[Test build #91773 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91773/testReport)** for PR 21529 at commit [`40abcff`](https://github.com/apache/spark/commit/40abcff647748719dc775fd3cbce09661323ca8b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3983/
Test PASSed.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/execution/PlannerSuite.scala:
```diff
@@ -679,6 +679,23 @@ class PlannerSuite extends SharedSQLContext {
+    outputPlan match {
+      case SortMergeJoinExec(leftKeys, rightKeys, _, _, _, _) =>
+        assert(leftKeys == Seq(exprA, exprA))
+        assert(rightKeys.contains(exprB) && rightKeys.contains(exprC))
```
is it better to check `rightKeys == Seq(exprB, exprC)`

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala:
```diff
@@ -882,4 +882,15 @@ class JoinSuite extends QueryTest with SharedSQLContext {
+      SQLConf.AUTO_BROADCASTJOIN_THRESHOLD.key -> "-1") {
+      val df1 = spark.range(0, 100, 1, 2)
+      val df2 = spark.range(100).select($"id".as("b1"), (- $"id").as("b2"))
+      val res = df1.join(df2, $"id" === $"b1" && $"id" === $"b2")
```
one difference between this test and the code in JIRA ticket is, the code in JIRA ticket has a Project above join, to trigger the double transformation issue. We should add a Project and make sure this test **does** fail without this patch.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/JoinSuite.scala:
```diff
@@ -882,4 +882,15 @@ class JoinSuite extends QueryTest with SharedSQLContext {
     }
   }
+
+  test("SPARK-24495: EnsureRequirements can return wrong plan when reusing the same key in join") {
```
as end-to-end test, maybe a better name is: `SPARK-24495: Join may return wrong result when having duplicated equal-join keys`

cloud-fan commented:
LGTM except some minor comments about test

SparkQA commented:
**[Test build #91773 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91773/testReport)** for PR 21529 at commit [`40abcff`](https://github.com/apache/spark/commit/40abcff647748719dc775fd3cbce09661323ca8b).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91773/
Test PASSed.

mgaido91 committed:
address comments

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/104/
Test PASSed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3995/
Test PASSed.

SparkQA commented:
**[Test build #91790 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91790/testReport)** for PR 21529 at commit [`6553c27`](https://github.com/apache/spark/commit/6553c27bc23efbaed268556b0cb19405fa693de9).

gatorsmile reviewed (commented):
LGTM

cloud-fan commented:
@mgaido91 can you fix the conflicts? thanks!

SparkQA commented:
**[Test build #91790 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91790/testReport)** for PR 21529 at commit [`6553c27`](https://github.com/apache/spark/commit/6553c27bc23efbaed268556b0cb19405fa693de9).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91790/
Test FAILed.

gatorsmile commented:
retest this please

SparkQA commented:
**[Test build #91797 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91797/testReport)** for PR 21529 at commit [`6553c27`](https://github.com/apache/spark/commit/6553c27bc23efbaed268556b0cb19405fa693de9).

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/3999/
Test PASSed.

AmplabJenkins commented:
Build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/109/
Test PASSed.

SparkQA commented:
**[Test build #91797 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91797/testReport)** for PR 21529 at commit [`6553c27`](https://github.com/apache/spark/commit/6553c27bc23efbaed268556b0cb19405fa693de9).
 * This patch **fails Spark unit tests**.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91797/
Test FAILed.

mgaido91 committed:
Merge branch 'master' of github.com:apache/spark into SPARK-24495

SparkQA commented:
**[Test build #91826 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91826/testReport)** for PR 21529 at commit [`6ef4f0d`](https://github.com/apache/spark/commit/6ef4f0df7590f0da5aa900f29292ec0fe94658fb).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/129/
Test PASSed.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution/4018/
Test PASSed.

SparkQA commented:
**[Test build #91826 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91826/testReport)** for PR 21529 at commit [`6ef4f0d`](https://github.com/apache/spark/commit/6ef4f0df7590f0da5aa900f29292ec0fe94658fb).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91826/
Test FAILed.

mgaido91 commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/142/
Test PASSed.

SparkQA commented:
**[Test build #91843 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/91843/testReport)** for PR 21529 at commit [`6ef4f0d`](https://github.com/apache/spark/commit/6ef4f0df7590f0da5aa900f29292ec0fe94658fb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/91843/
Test PASSed.

gatorsmile commented:
Thanks! Merged to master/2.3

mgaido91 commented:
Thanks @gatorsmile. Sorry, may I ask what you think about https://github.com/apache/spark/pull/21529#issuecomment-396707622? Thanks.

gatorsmile commented:
Adding new queries to `SQLQueryTestSuite` is the best way to do it in the current infrastructure. Do your best to cover all the join algorithms for different input data and join types?
